Centralize install code

lavalamp opened the pull request:
Trying to figure out a way to do this that makes the changes as painless to roll out as possible. This is going to be a multi-step process...

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30565)

<!-- Reviewable:end -->

lavalamp committed:
Factor out InterfacesFor

lavalamp committed:
Factor out the horrible install code that everyone copies

lavalamp committed:
Update batch to use new install code

lavalamp committed:
Add warning comment

lavalamp commented:
OK, I think this is ready for review. I've structured this so that I can refactor in smallish chunks.

lavalamp commented:
First commit is #30663.

smarterclayton commented:
I knew it was only a matter of time before we got factories.

nikhiljindal commented:
I assume this is not for 1.4, but want to confirm

lavalamp commented:
Factory produced groups are better than artisanal hand-crafted groups!

@nikhiljindal it would be great to get in. It should be pretty easy to review at this point? Hopefully the tests pass now.

lavalamp commented:
OK, I think I fixed the boilerplate & test for real this time.

nikhiljindal commented on pkg/api/testapi/testapi.go:
```diff
@@ -15,6 +15,12 @@ limitations under the License.
+// TODO(lavalamp): this package is a huge disaster at the moment. I intend to
+// refactor. All code currently using this package should change:
+// 1. Declare your own registered.APIGroupRegistrationManager in your own test code.
+// 2. Import the relevant instal packages.
```
install

nikhiljindal commented on pkg/apimachinery/announced/announced.go:
```diff
@@ -0,0 +1,100 @@
+// distinct from registration (in the 'registered' package) in that it's safe
+// to announce every possible group linked in, but only groups requested at
+// runtime should be registered. This package contains both a registry, and
+// factory code (which was formerly copy-pasta in every install package).
```
pasta -> paste

lavalamp replied:
This is actually deliberate-- copy-pasta is for code that's been copied and pasted and is now spaghetti ;)

if you google "copypasta code" you'll see that I'm not the first to use it this way.

nikhiljindal commented on pkg/apimachinery/announced/announced.go:
```diff
@@ -0,0 +1,100 @@
+
+// Package announced contains tools for announcing API group factories. This is
+// distinct from registration (in the 'registered' package) in that it's safe
+// to announce every possible group linked in, but only groups requested at
```
Isnt that the difference between registered and enabled? You can register all groups, but group will be added to Scheme only when it is enabled.

lavalamp replied:
Added a section here and in the registered package.

nikhiljindal commented on pkg/apimachinery/announced/announced.go:
```diff
@@ -0,0 +1,100 @@
+// (e.g., for testing).
+type APIGroupAnnouncementRegistry map[string]*GroupMetaFactory
+
+func (gfr APIGroupAnnouncementRegistry) group(groupName string) *GroupMetaFactory {
```
Why gfr? Should it be gar?

nikhiljindal commented on pkg/apimachinery/announced/group_factory.go:
```diff
@@ -0,0 +1,244 @@
+)
+
+type ApplyToScheme func(*runtime.Scheme) error
+type VersionToSchemeFunc map[string]ApplyToScheme
```
its not a func?

lavalamp replied:
Parse as "Version To SchemeFunc"-- I'll change the name of ApplyToScheme to make this clearer

nikhiljindal commented on pkg/apimachinery/announced/group_factory.go:
```diff
@@ -0,0 +1,244 @@
+	RootScopedKinds sets.String // nil is allowed
+	IgnoredKinds    sets.String // nil is allowed
+
+	// May be nil if none
```
incomplete sentence?

nikhiljindal commented on pkg/apimachinery/announced/group_factory.go:
```diff
@@ -0,0 +1,244 @@
+		)
+	}
+
+	availableVersions := []unversioned.GroupVersion{}
```
Is unprioritizedVersions a better name.
Whatever the name, add a comment saying this is the array of versions that are registered, but are not in the prioritized versions list.

nikhiljindal commented on pkg/apimachinery/announced/group_factory.go:
```diff
@@ -0,0 +1,244 @@
+		availableVersions = append(availableVersions, unversioned.GroupVersion{v.GroupName, v.VersionName})
+	}
+	if len(availableVersions) != 0 {
+		glog.Warningf("group %v has unprioritized versions %#v. They will have an arbitrary preference order!", gmf.GroupArgs.GroupName, availableVersions)
```
Error?

lavalamp replied:
Hm, maybe? I guess I'd rather wait and see what people end up doing with this.

lavalamp replied:
Actually it's only problematic if there's more than one of these, now that I think about it :)

nikhiljindal commented:
There is a lot of complexity here: announced, registered and enabled group versions; factory for announcing, registering and enabling group versions; a registry of such factories; APIRegistrationManager.
Its not easy to wrap your head around all of this.

Do we need all of these layers? Any layer that we dont really need right now?

nikhiljindal commented:
cc @kubernetes/sig-api-machinery

lavalamp commented:
Yeah, this is complex but the complexity is already there. This makes it
centralized complexity instead of distributed complexity. It can still be
significantly improved but I want to make the improvements in 2 packages
and not 20.

On Mon, Aug 22, 2016 at 11:49 PM, Nikhil Jindal <notifications@github.com>
wrote:

> There is a lot of complexity here: announced, registered and enabled group
> versions; factory for announcing, registering and enabling group versions;
> a registry of such factories; APIRegistrationManager.
> Its not easy to wrap your head around all of this.
>
> Do we need all of these layers? Any layer that we dont really need right
> now?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30565#issuecomment-241642446>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglsSFcL6wiYm7lQTGlE97Y7_uHQwZks5qipf3gaJpZM4Jjkrl>
> .
>

nikhiljindal commented:
> Yeah, this is complex but the complexity is already there. This makes it
> centralized complexity instead of distributed complexity. 

Yes the complexity was still there. But now there are new concepts: announced group versions, registery, factory.

> It can still be
> significantly improved but I want to make the improvements in 2 packages
> and not 20.

Agreed, thats reasonable.


A few points:
* GroupMetaFactory does not sound like the right name: With that name, I expect it to produce a GroupMeta, but what it does is register and enable the right group versions in APIRegistrationManager. Should it be APIRegistrationManager{Generator, Builder, Factory}?
* Announce: All it does is add a group version to the registry. How about calling it Add instead of announce? 

Without having worked out exact details, a high level flow that makes sense is: you *Add* your group versions to a APIRegistrationManagerFactory, the factory produces APIRegistrationManager which knows which group versions and resources are enabled and then all other code just uses APIRegistrationManager.
Right now, we have code in install.go on how to update APIRegistrationManager. With this change, they will just add their group version to the factory, which has the common code to update APIRegistrationManager.

lavalamp commented:
here's two commits with the first batch of fixes. Still pondering your name suggestions.

lavalamp commented:
re: naming

Today, the API group registration manager has two concepts:
* Registration, which I want to start calling "announcing" and move to the announcement manager in the new package.
* Enabling, which I want to leave in the registered package, and start calling "registration".

If you look at GroupMetaFactory, it does actually produce GroupMeta objects, but it's tightly coupled with the registration manager. The idea I'm trying to present here is that you "announce" your group's components, those get combined into a GroupMetaFactory, and then anyone can use the factory to place the correct GroupMeta in their APIGroupRegistrationManager.

lavalamp commented:
OK, I changed the name to "APIGroupFactoryRegistry" and added comments in several places justifying the "GroupMetaFactory" name. Added other misc comments. Squashed.

Thanks for the comments, can you live with it this way until the next installment? :)

nikhiljindal commented:
LGTM, thanks

pwittrock commented:
Added do not merge label to sync up with Daniel on this before it goes in

luxas commented:
Please say this is going to make v1.4...

pwittrock commented:
Marking this as P0 because rebasing it has created a lot of work for Daniel and he has been trying to get it in since pre-1.3, but pulled off of it to perform heroics on the release.  Moving it to milestone v1.5.

pwittrock commented:
@luxas I talked with Daniel and I would like this to be the first PR for v1.5.  I don't believe putting it in v1.4 serves the interest of the release.  I know realize this is super painful and thank you for your patience.  Msg me on slack if you want to chat further about this decision.

lavalamp commented:
(Toggled the labels--forgot to add that milestone to the blacklist. Fixed now.)

lavalamp commented:
@k8s-bot test this issue #IGNORE
